src/framework/audio/internal/rpc/irpcchannel.h

```diff
@@ -0,0 +1,352 @@
+}
+
+struct Msg {
+    CallId callId = 0;
+    Method method = Method::Undefined;
+    MsgType type = MsgType::Undefined;
+    ByteArray data;
+};
+
+using Handler = std::function<void (const Msg& msg)>;
+
+// stream
+using StreamId = uint32_t;
+inline StreamId new_stream_id()
+{
+    static StreamId lastId = 0;
+    ++lastId;
+    return lastId;
+}
+
+struct StreamMsg {
+    StreamId streamId = 0;
+    ByteArray data;
+};
+
+using StreamHandler = std::function<void (const StreamMsg& msg)>;
+
+enum class StreamType {
+    Undefined = 0,
+    Send,
+    Receive
+};
+
+struct IRpcStream {
+    virtual ~IRpcStream() = default;
+
+    virtual StreamId streamId() const = 0;
+    virtual StreamType type() const = 0;
+    virtual void init() = 0;
+    virtual bool inited() const = 0;
+};
+
+class IRpcChannel;
+template<typename ... Types>
+class RpcStream : public IRpcStream, public async::Asyncable
+{
+public:
+    RpcStream(IRpcChannel* rpc, StreamId id, StreamType type, const async::Channel<Types...>& ch)
+        : m_rpc(rpc), m_streamId(id), m_type(type), m_ch(ch) {}
+
+    StreamId streamId() const override { return m_streamId; }
+    StreamType type() const override { return m_type; }
+    void init() override;
+    bool inited() const override { return m_inited; }
+
+private:
+    IRpcChannel* m_rpc = nullptr;
+    StreamId m_streamId = 0;
+    StreamType m_type = StreamType::Undefined;
+    async::Channel<Types...> m_ch;
+    bool m_inited = false;
+};
+
+class IRpcChannel : MODULE_EXPORT_INTERFACE
+{
+    INTERFACE_ID(IRpcChannel)
+public:
+    virtual ~IRpcChannel() = default;
+
+    virtual void send(const Msg& msg, const Handler& onResponse = nullptr) = 0;
+    virtual void onMethod(Method method, Handler h) = 0;
+    virtual void listenAll(Handler h) = 0;
+
+    // stream (async/channel)
+    template<typename ... Types>
+    StreamId addSendStream(const async::Channel<Types...>& ch)
+    {
+        StreamId id = new_stream_id();
+        std::shared_ptr<IRpcStream> s = std::shared_ptr<IRpcStream>(new RpcStream<Types...>(this, id, StreamType::Send, ch));
+        addStream(s);
+        return id;
+    }
+
+    template<typename ... Types>
+    void addReceiveStream(rpc::StreamId id, const async::Channel<Types...>& ch)
+    {
+        std::shared_ptr<IRpcStream> s = std::shared_ptr<IRpcStream>(new RpcStream<Types...>(this, id, StreamType::Receive, ch));
+        addStream(s);
+    }
+
+    virtual void addStream(std::shared_ptr<IRpcStream> s) = 0;
+    virtual void removeStream(StreamId id) = 0;
+    virtual void sendStream(const StreamMsg& msg) = 0;
+    virtual void onStream(StreamId id, StreamHandler h) = 0;
+};
+
+template<typename ... Types>
+void RpcStream<Types...>::init()
+{
+    if (m_inited) {
+        return;
+    }
+
+    switch (m_type) {
+    case StreamType::Send: {
+        m_ch.onReceive(this, [this](const Types... args) {
+                ByteArray data = RpcPacker::pack(args ...);
```

Comment:
Can we use std::move here?

Reply:
no point, ByteArray has implicit copy, data is not copied